### PR TITLE
Set the order to "failed" when a dispute arrives

### PR DESCRIPTION
**Pull request by DanReyLop**
See discussion in p1574360413049100-slack-team-hydra-stripe

### The problem

- SEPA payments are asynchronous.
- If the payment is declined for some reason, it's treated as a disputed by Stripe.
- There's no appeal process, disputes are immediately closed in favor of the customer.
- When a dispute is opened, the order is set to `on-hold`, indefinitely, which is not ideal.

### The solution

Instead of setting the order to `on-hold`, set it to `failed`. Can y'all think of any reason this is a bad idea? We aren't listening to the `charge.dispute.closed` webhook, so those orders (even the credit card ones) would be set to `on-hold` forever. I didn't want to spend a lot of time on this anyway, since this will be revamped when the order status system changes.

### Alternatives

We could start listening to the `charge.dispute.closed` event. That event would tell us if the dispute has been won or lost. But... then what? We add a nice order note and set the order to... what? I can't think on any status transition we could make that would be correct all the time.

### Steps to reproduce

- Use a Stripe account that has SEPA payments enabled. I used the one from the Secret Store.
- Configure webhooks to point to your site on that account.
- Enable the SEPA payment method.
- Make a payment using the `DE35370400440532013002` account number (that will immediately create a dispute).
- Go to the `Edit Order` screen, you'll see that the order status is `Failed`.

cc/ @jrick1229 @RadoslavGeorgiev

### CODE REVIEW

File: includes/class-wc-stripe-webhook-handler.php [hunk 1]
```diff
@@ -292,13 +292,16 @@ public function process_webhook_dispute( $notification ) {
 			return;
 		}
 
+		// Sometimes (at least in testing) the "dispute.created" and "charge.succeeded" events can arrive really close for SEPA payments.
+		// Use the locking system to prevent the charge.succeeded to set the order to "processing" even after the dispute was created.
+		$this->lock_order_payment( $order );
```
Comment (vbelolapotkov) [inline includes/class-wc-stripe-webhook-handler.php]: Great idea! 👍

File: includes/class-wc-stripe-webhook-handler.php [hunk 2]
```diff
@@ -292,13 +292,16 @@ public function process_webhook_dispute( $notification ) {
...
+
 		/* translators: 1) The URL to the order. */
-		$order->update_status( 'on-hold', sprintf( __( 'A dispute was created for this order. Response is needed. Please go to your <a href="%s" title="Stripe Dashboard" target="_blank">Stripe Dashboard</a> to review this dispute.', 'woocommerce-gateway-stripe' ), $this->get_transaction_url( $order ) ) );
+		$order->update_status( 'failed', sprintf( __( 'A dispute was created for this order. Response is needed. Please go to your <a href="%s" title="Stripe Dashboard" target="_blank">Stripe Dashboard</a> to review this dispute.', 'woocommerce-gateway-stripe' ), $this->get_transaction_url( $order ) ) );
 
 		do_action( 'wc_gateway_stripe_process_webhook_payment_error', $order, $notification );
 
-		$order_id = WC_Stripe_Helper::is_wc_lt( '3.0' ) ? $order->id : $order->get_id();
-		$this->send_failed_order_email( $order_id );
```
Comment (vbelolapotkov) [inline includes/class-wc-stripe-webhook-handler.php]: As I see those lines would send an email about failed order. Why do we change that?

File: includes/class-wc-stripe-webhook-handler.php [hunk 3]
```diff
@@ -380,6 +383,10 @@ public function process_webhook_charge_succeeded( $notification ) {
 			return;
 		}
 
+		if ( $this->lock_order_payment( $order ) ) {
+			return;
+		}
+
 		$order_id = WC_Stripe_Helper::is_wc_lt( '3.0' ) ? $order->id : $order->get_id();
 
 		if ( ! $order->has_status( 'on-hold' ) ) {
```
Comment (vbelolapotkov) [inline includes/class-wc-stripe-webhook-handler.php]: @DanReyLop does it make sense adding `failed` status to this check?

File: includes/class-wc-stripe-webhook-handler.php (hunk 2 above)
Comment (DanReyLop) [reply includes/class-wc-stripe-webhook-handler.php]: When the order is set to `failed`, the "Order Failed" email is already triggered automatically, so these lines are redundant (in fact, they would send another email).

**Comment by DanReyLop** [comment]
@vbelolapotkov Thanks for the review! Unfortunately, this PR doesn't really fix the issue. @RadoslavGeorgiev reviewed it a long time ago and we discussed (mostly he discussed and I nodded) how to properly fix it. The problem is that Stripe will send the site several webhooks at the same-ish time when a dispute is created, so dealing with duplicated actions and the order of the actions is way trickier than it should.

@RadoslavGeorgiev Did you get around to this? If not, I can open a new issue with a more detailed description of how to solve the bug, and try to fix it for the next release.

**Comment by vbelolapotkov** [comment]
@DanReyLop thanks for the update! Would be happy to hear @RadoslavGeorgiev 's thoughts too 🙂
There is an issue #1098 for adding other webhooks. Do you think implementing it will fix the current one as well?

**Comment by DanReyLop** [comment]
> There is an issue #1098 for adding other webhooks. Do you think implementing it will fix the current one as well?

@vbelolapotkov This PR focuses on SEPA disputes, which are immediately **accepted**. We are already handling the "dispute opened" webhook. The problem is that many webhooks are triggered in rapid succession ("dispute opened", "dispute accepted", "funds withdrawn", etc) and ordering is not guaranteed, so it gets tricky. #1098 seems like a completely different issue, basically it would be adding new functionality by listening to a webhook that we are currently not listening to.

**Comment by vbelolapotkov** [comment]
> The problem is that many webhooks are triggered in rapid succession ("dispute opened", "dispute accepted", "funds withdrawn", etc) and ordering is not guaranteed, so it gets tricky.

@DanReyLop, understood. You've mentioned above that you discussed a proper way to fix it with @RadoslavGeorgiev. Do you have a reference for the discussion or solution?

**Comment by dechov** [comment]
@DanReyLop So is the problem that `charge.succeeded` can arrive after `charge.dispute.created` (or _right_ before it and lock) and it would appear to succeed instead of fail?

In that case, could we _not_ lock the charge.succeeded processing, and also encode the disputed state somehow so that it doesn't process charge.succeeded in that case?

**Comment by DanReyLop** [comment]
> So is the problem that charge.succeeded can arrive after charge.dispute.created (or right before it and lock) and it would appear to succeed instead of fail?

I opened this PR half a year ago, and had a conversation about it with @RadoslavGeorgiev 3 months ago. I don't remember *any* of this.

I just remember thinking that the whole webhook processing system is a house of cards, because Stripe doesn't guarantee the order it sends them, and PHP's execution model doesn't exactly prevent race conditions either.

I'm just gonna close this PR, it doesn't work in all cases so it's not helping anyone. Long-term, I think we should somehow implement a queue or a global locking mechanism so webhooks only get processed one by one. Maybe we could leverage Action Scheduler. That wouldn't solve the problem of Stripe sending webhooks in an unexpected order, but it would help.

File: includes/class-wc-stripe-webhook-handler.php (hunk 2 above)
Comment (dechov) [reply includes/class-wc-stripe-webhook-handler.php:301]: If that's true, then extra emails are already being sent in certain other cases, e.g.:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/19e5f4a47d4ad94d4d367a588b3cd4b227c9f80d/includes/class-wc-stripe-webhook-handler.php#L673-L678

Should those `send_failed_order_email` calls be removed? Can open an issue to look into that.

Comment (DanReyLop) [reply includes/class-wc-stripe-webhook-handler.php:301]: > Should those send_failed_order_email calls be removed?

I think they should, yes.

Comment (dechov) [reply includes/class-wc-stripe-webhook-handler.php:301]: 👍 Opened https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1229

**Comment by zwiesman31** [comment]
@shendy-a8c & @v18 Any updates here?

**Pull request closed without merging**